[Tiling rewrite pt1] Normalize reads and writes to common iter space

eellison opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #154768
* #153751
* #153748
* #153730
* __->__ #153723

In order to take the globally best tiling, we need to normalize all the node read and writes to a common iteration space. This first pr finds a common split among nodes in a fused scheduler node, and then normalizes reads and writes to the common split. 


cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @chenyang78 @kadeng @muchulee8 @amjames @chauhang @aakhundov

eellison committed:
Update

[ghstack-poisoned]

eellison committed:
Update

[ghstack-poisoned]

eellison committed:
Update

[ghstack-poisoned]

etaf commented on test/inductor/test_loop_ordering.py:
```diff
@@ -507,6 +517,231 @@ def f(x):
+            def foo(x, y):
+                return x + y
+
+            foo(torch.rand([4, 4], device="cuda"), torch.rand([4, 4], device="cuda").T)
```
Hi, May I suggest we mark these cases as requires_cuda or replace the hardcode `cuda` with GPU_TYPE here? These new test case will also run on XPU and fail with `cuda`, thanks.

jansel reviewed (changes requested):
Test failures?

eellison committed:
Update on "[Tiling rewrite pt1] Normalize reads and writes to common iter space"


In order to take the globally best tiling, we need to normalize all the node read and writes to a common iteration space. This first pr finds a common split among nodes in a fused scheduler node, and then normalizes reads and writes to the common split. 


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

jansel commented on torch/_inductor/codegen/simd.py:
```diff
@@ -705,21 +705,33 @@ def getter(flat_vars: list[sympy.Expr]) -> sympy.Expr:
         sizevars = V.graph.sizevars
         if len(lengths[1]) == 0 and (
-            sizevars.statically_known_equals(
+            not sizevars.is_expr_static_and_true(reduction_numel == sympy.S.One)
```
Doesn't `==` for sympy.Expr always return True/False?  It does something different than `sympy.Eq()`

eellison committed:
Update on "[Tiling rewrite pt1] Normalize reads and writes to common iter space"


In order to take the globally best tiling, we need to normalize all the node read and writes to a common iteration space. This first pr finds a common split among nodes in a fused scheduler node, and then normalizes reads and writes to the common split. 


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

eellison committed:
Update

[ghstack-poisoned]

eellison committed:
Update

[ghstack-poisoned]

eellison committed:
Update

[ghstack-poisoned]

eellison committed:
Update

[ghstack-poisoned]
